File: src/easy/LowestCommonAncestor.java
package easy;

import common.TreeNode;

// https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/
public class LowestCommonAncestor {
  public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
    if (root == null || p == root || q == root) return root;
    TreeNode l = lowestCommonAncestor(root.left, p, q);
    TreeNode r = lowestCommonAncestor(root.right, p, q);
    return l == null ? r : r == null ? l : root;
  }
}

File: csharp/LeetCode/Easy/LowestCommonAncestor.cs
namespace LeetCode.Easy.LowestCommonAncestor
{
    using LeetCode.DataStructures;

    public class Solution
    {
        public TreeNode LowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q)
        {
            while ((root.val - p.val) * (root.val - q.val) > 0)
            {
                root = root.val > p.val ? root.left : root.right;
            }
            return root;
        }
    }
}
